### ISSUE TITLE
VS 16.7.0 P3.1 Android SDK License Acceptance Dialog

### ISSUE THREAD

**Issue by gmck** [xamarin/xamarin-android#4880]
Nearly every compiler error triggers this dialog. When I Accept, the build window then displays the following.

```
No Android SDK instance matching the 'C:\Program Files (x86)\Android\android-sdk' path
Element 'http://schemas.android.com/repository/android/sites-common/1:site-list' has a duplicate attribute 'http://www.w3.org/2000/xmlns/:common' in document 'https://dl.google.com/android/repository/addons_list-3.xml'. New value will override the old one
Element 'http://schemas.android.com/repository/android/sites-common/1:site-list' has a duplicate attribute 'http://www.w3.org/2000/xmlns/:sdk' in document 'https://dl.google.com/android/repository/addons_list-3.xml'. New value will override the old one
Element 'http://schemas.android.com/sdk/android/repo/addon2/01:sdk-addon' has a duplicate attribute 'http://www.w3.org/2000/xmlns/:sdk' in document 'https://dl.google.com/android/repository/addon2-1.xml'. New value will override the old one
Package loaded from https://dl.google.com/android/repository/extras/intel/addon2-1.xml [603:3], with path 'extras;intel;Hardware_Accelerated_Execution_Manager' has no valid archives (Host OS: Windows)
Package loaded from https://dl.google.com/android/repository/extras/intel/addon2-1.xml [603:3], with path 'extras;intel;Hardware_Accelerated_Execution_Manager' has no valid archives (Host OS: Windows)
Android component "tools" is obsolete, please install "cmdline-tools"
Trying to fallback to "tools"...
File 'C:\Program Files (x86)\Android\android-sdk\cmdline-tools\android.bat' not found for 'Android SDK Core'
File 'C:\Program Files (x86)\Android\android-sdk\cmdline-tools\mksdcard.exe' not found for 'Android SDK Core'
File 'C:\Program Files (x86)\Android\android-sdk\cmdline-tools\emulator.exe' not found for 'Android SDK Core'
Installation completed
```
The first time it happened, after updating to Preview 3.1, I then just updated the SDK manually. All is ok until the next compiler error when the dialog reappears. If I run it again - same result. Again all is ok until the next compiler error and back it comes. The **cmdline-tools** folder was created 24/07/2020 and contains a subfolder **latest**, which contains **bin** and **lib** folders both dated 24/07/2020. There are no files in cmdline-tools folder only the folder latest, so the _Trying to fallback to "tools".._ errors would be correct.

Manually running the Android SDK Manager within VS indicates there are no available updates.

**Comment by brendanzagaeski** [comment]
Thanks for submitting this item! This is an unfortunate known issue in Visual Studio 2019 version 16.7 Preview 3, but there are a few workaround options available:

- **Option A –** Set the `AndroidCommandLineToolsVersion` MSBuild property to `latest`:

  1. Open the _.csproj_ project file for the Xamarin.Android app project [in Visual Studio](https://docs.microsoft.com/visualstudio/msbuild/visual-studio-integration-msbuild?view=vs-2019#edit-project-files-in-visual-studio) or another text editor.

  2. Find an existing closing `</PropertyGroup>` tag in the file, and add the following additional `<PropertyGroup>` element after it:

     ```xml
     <PropertyGroup>
       <AndroidCommandLineToolsVersion>latest</AndroidCommandLineToolsVersion>
     </PropertyGroup>
     ```

  This option is included in the Visual Studio release notes, but it's a bit hard to find: <https://docs.microsoft.com/visualstudio/releases/2019/release-notes-preview#xamarin>

  After a fix for this issue becomes available in the next Visual Studio 2019 version 16.7 Preview, be sure to remove the workaround `<PropertyGroup>` element to allow the project to use the default recommended value of `AndroidCommandLineToolsVersion`.

- **Option B –** Temporarily disable the **Auto Install Android SDKs** feature and ignore the message in the **Error List**:

  Open **Tools > Options**, open the **Xamarin > Android Settings** item, and uncheck **Auto Install Android SDKs**.

  After a fix for this issue becomes available in the next Visual Studio 2019 version 16.7 Preview, be sure to re-enable **Auto Install Android SDKs**.

- **Option C –** Click **Cancel** on the **Android SDK - License Acceptance** dialog each time it pops up and ignore the message in the **Error List**.

  Explanation: In this particular case, the "missing Android SDKs" error does not interfere with the build process, so it is OK to ignore the error and dismiss the **Android SDK - License Acceptance** dialog each time it appears.

  The inconvenience with this approach is, as you noticed, that the dialog will appear again each time a Xamarin.Android app project is opened and each time a Xamarin.Android app build has an error.

**Comment by brendanzagaeski** [comment]
_Cross-referencing note for the Xamarin team_

Corresponding Developer Community item: <https://developercommunity.visualstudio.com/content/problem/1092334/vs2019-167-preview-3031-the-project-is-missing-and.html>

**Comment by gmck** [comment]
@brendanzagaeski  Thanks for that. Latest fixed it.

Also in that link, I noticed a note re: 

`Code fixers have been introduced for the most common types of problems experienced by users with Android layout files.`

Was that meant to address the problems I outlined in #4317? Unfortunately, I still can't open any xml layouts using the android designer in this project. Referenced in email to you and @dellis1972  13/06/2020.

**Comment by brendanzagaeski** [comment]
Glad that worked for the Android SDK licenses dialog issue!

> Also in that link, I noticed a note re:
> 
> `Code fixers have been introduced for the most common types of problems experienced by users with Android layout files.`
> 
> Was that meant to address the problems I outlined in #4317?

Ah no, I think "code fixers" would more accurately be termed "code fix Quick Actions" to match the [documentation on Quick Actions](https://docs.microsoft.com/visualstudio/ide/quick-actions?view=vs-2019). I'll plan to check on the authorship of that note and suggest the wording change.

**Comment by donaldashworth** [comment]
This work-around did not solve the issue for me.  Receiving "No Android SDK instance matching the 'C:\Program Files (x86)\Android\android-sdk' path" after upgrading to Microsoft Visual Studio Professional 2019 Preview
Version 16.7.0 Preview 3.1 as recommended by the Xamarin team to correct "build hang issues".

Now I cannot compile our Xamarin mobile app of which has been published in both App Center and Google Play.  In sort, I am (once again) dead in the water.

Any other suggestions?

**Comment by brendanzagaeski** [comment]
@donaldashworth, if you get a chance, I'd recommend submitting a new [Visual Studio Developer Community](https://developercommunity.visualstudio.com/spaces/8/index.html/) feedback item via [**Help > Send Feedback > Report a Problem**](https://docs.microsoft.com/visualstudio/ide/feedback-options?view=vs-2019) in Visual Studio for the behavior you are seeing, and if possible, allow the tool to attach the default logs that it collects. That will provide a good starting place for the team to help further diagnose what might be happening in your environment. Thanks!

Additional background info: The workaround for this item will only help if the **Android SDK - License Acceptance** dialog shows **Android SDK Command-line Tools** as the single item that's prompting for license acceptance:

![CommandLineToolsAcceptance](https://user-images.githubusercontent.com/4934473/86964904-9e293580-c11b-11ea-9e88-43ceb55954b9.png)

As a first guess, it sounds like there might be another complication involved in your environment

**Comment by donaldashworth** [comment]
Yup.  The Android Sdk License issue is what results in the Verbose Xamarin output window "No Android SDK instance matching the 'C:\Program Files (x86)\Android\android-sdk' path" message.  I have already submitted a trouble report...was hoping there may be another way to get around this such that I can do my job.

**Comment by brendanzagaeski** [comment]
> I have already submitted a trouble report...was hoping there may be another way to get around this such that I can do my job.

@donaldashworth, ah perfect. I searched a little just now to see if I could find that report to take a look at the logs, but I couldn't find it. I probably didn't have quite the right search query. If you like, feel free to reply here with the link to the Visual Developer Community item here, and I can take a look.

The particular symptom where the workaround helps is also in theory ignorable (workaround Option C) because it doesn't prevent building successfully, so if your build process is actually getting blocked right now, then it sounds like there are other details about your specific scenario that are adding another layer of complexity.

**Comment by JayD2019** [comment]
I have this same issue in VS2019  v16.8.4. (on Windows 7 SP1) 
The license acceptance window suddenly started popping up yesterday (Feb 1 2021 ). The SDK manager is "empty" ; nothing listed in either tab. 

As per workaround above, I have unchecked the auto-install Android SDK - for now - and am ignoring the error. Solution will build, but DDMS will not open, nor will Android Device Manager.

**Comment by donaldashworth** [comment]
We’ve pretty much given up on this issue.

**Comment by JayD2019** [comment]
Hi Donald,

With or without the  auto-install Android SDK checked, I cannot get the Android SDK to update - I just keep getting the license message and I go around in a loop. Even in a newly created project (see screenshots )  
[AndroidSDKerror_in VS2019.docx](https://github.com/xamarin/xamarin-android/files/5968766/AndroidSDKerror_in.VS2019.docx)

I open the SDK Manager and BOTH tabs are empty , when I try to update from the repository nothing happens and
in the output window I get this error.
_No Android SDK instance matching the 'C:\Program Files (x86)\Android\android-sdk' path
Unable to perform detection in 'Xamarin Repository', no known components
Android component "tools" is obsolete, please install "cmdline-tools"
Trying to fallback to "tools"...
Could not retrieve Pkg.Revision from C:\Program Files (x86)\Android\android-sdk\tools\source.properties, r0 assumed
Installation completed_

Also, DDMS will not open nor will Android Device Manager. 

In Tools>Options> Xamarin,    C:\Program Files (x86)\Android\android-sdk is the correct location 
![image](https://user-images.githubusercontent.com/46284200/107706457-b86d5800-6c96-11eb-9f5a-7f1fc474166b.png)

Also, I have "cmdline-tools" installed as C:\Program Files (x86)\Android\android-sdk\cmdline-tools

Is this a new issue or the same?  I am relatively new to Xamarin.Android Platform. 
It all seems to be related to the Android License error.

Thanks ,
J

**Comment by JayD2019** [comment]
I have finally fixed my issue. I had to uninstall the Android Tools installation - and - delete everything from  C:\Program Files (x86)\Android\android-sdk\ .   Uninstall Visual Studio and then re-install it (repair did not work) . Then I was able to open and re-populate the SDK manager and open all the other android tools. Archive All now works also.  Also the Android license manager behaved normally also. I know of at least one other person who has this same issue in VS2019 v16,8.5 , so I hope this may help anyone else out there. 
Ciao 
J
